app/src/org/commcare/android/logging/ForceCloseLogger.java

```diff
@@ -26,61 +26,65 @@
 
 /**
  * Catch exceptions that are going to crash the phone, grab the stack trace,
- * and upload to developers.
+ * and upload to device logs.
  *
  * @author csims@dimagi.com
  **/
-public class ExceptionReporting {
-    private static final String TAG = ExceptionReporting.class.getSimpleName();
+public class ForceCloseLogger {
+    private static final String TAG = ForceCloseLogger.class.getSimpleName();
+
+    private static SqlStorage<ForceCloseLogEntry> logStorage;
+
+    public static void registerStorage(SqlStorage<ForceCloseLogEntry> storage) {
+        logStorage = storage;
+    }
 
     public static void reportExceptionInBg(final Throwable exception) {
         new Thread(new Runnable() {
             public void run() {
-                sendExceptionToServer(exception);
+                sendToServerOrStore(exception);
             }
         }).start();
     }
 
-    private static void sendExceptionToServer(Throwable exception) {
-        ByteArrayOutputStream baos = new ByteArrayOutputStream();
-
-        //TODO: This is ridiculous. Just do the normal log submission process
-        DeviceReportWriter report;
-        try {
-            report = new DeviceReportWriter(baos);
-        } catch (IOException e) {
-            report = null;
-        }
-
+    /**
+     * Attempts to send a force close report for the given exception to the server
+     * immediately. If this fails, instead writes the exception to storage as a ForceCloseLogEntry,
+     * so that we can attempt to send it again later during a normal LogSubmissionTask
+     */
+    private static void sendToServerOrStore(Throwable exception) {
+        ByteArrayOutputStream streamToWriteErrorTo = new ByteArrayOutputStream();
```

Comment:
Errr, unless the `DeviceReportWriter` constructor throws an `IOException`, then it isn't closed...